File: ios/Sources/RoadSenseNSBootstrap/Pipeline/ReadingWindowProcessor.swift
import Foundation

public struct PersistedReadingCandidate: Equatable, Sendable {
    public let latitude: Double
    public let longitude: Double
    public let roughnessRMS: Double
    public let speedKmh: Double
    public let headingDegrees: Double
    public let gpsAccuracyMeters: Double
    public let isPothole: Bool
    public let potholeMagnitudeG: Double?
    public let recordedAt: Date
    public let durationSeconds: TimeInterval

    public init(
        latitude: Double,
        longitude: Double,
        roughnessRMS: Double,
        speedKmh: Double,
        headingDegrees: Double,
        gpsAccuracyMeters: Double,
        isPothole: Bool,
        potholeMagnitudeG: Double?,
        recordedAt: Date,
        durationSeconds: TimeInterval
    ) {
        self.latitude = latitude
        self.longitude = longitude
        self.roughnessRMS = roughnessRMS
        self.speedKmh = speedKmh
        self.headingDegrees = headingDegrees
        self.gpsAccuracyMeters = gpsAccuracyMeters
        self.isPothole = isPothole
        self.potholeMagnitudeG = potholeMagnitudeG
        self.recordedAt = recordedAt
        self.durationSeconds = durationSeconds
    }
}

public enum ReadingWindowProcessingOutcome: Equatable, Sendable {
    case accepted(PersistedReadingCandidate)
    case rejected(QualityRejectionReason)
}

public enum ReadingWindowProcessor {
    public static func process(
        window: ReadingWindow,
        deviceState: DeviceCollectionState,
        potholeCandidates: [PotholeCandidate]
    ) -> ReadingWindowProcessingOutcome {
        switch QualityFilter.evaluate(reading: window, deviceState: deviceState) {
        case .accepted:
            return .accepted(
                PersistedReadingCandidate(
                    latitude: window.latitude,
                    longitude: window.longitude,
                    roughnessRMS: window.roughnessRMS,
                    speedKmh: window.speedKmh,
                    headingDegrees: window.headingDegrees,
                    gpsAccuracyMeters: window.gpsAccuracyMeters,
                    isPothole: !potholeCandidates.isEmpty,
                    potholeMagnitudeG: potholeCandidates.map(\.magnitudeG).max(),
                    recordedAt: Date(timeIntervalSince1970: window.startedAt + window.durationSeconds),
                    durationSeconds: window.durationSeconds
                )
            )
        case let .rejected(reason):
            return .rejected(reason)
        }
    }
}

File: android/core-sensor/src/main/kotlin/ca/roadsense/ns/sensor/ReadingWindowProcessor.kt
package ca.roadsense.ns.sensor

data class PersistedReadingCandidate(
    val latitude: Double,
    val longitude: Double,
    val roughnessRMS: Double,
    val speedKmh: Double,
    val headingDegrees: Double,
    val gpsAccuracyMeters: Double,
    val isPothole: Boolean,
    val potholeMagnitudeG: Double?,
    val recordedAtEpochSeconds: Double,
    val durationSeconds: Double,
)

sealed class ReadingWindowProcessingOutcome {
    data class Accepted(val candidate: PersistedReadingCandidate) : ReadingWindowProcessingOutcome()
    data class Rejected(val reason: QualityRejectionReason) : ReadingWindowProcessingOutcome()
}

object ReadingWindowProcessor {
    fun process(
        window: ReadingWindow,
        deviceState: DeviceCollectionState,
        potholeCandidates: List<PotholeCandidate>,
    ): ReadingWindowProcessingOutcome {
        return when (val decision = QualityFilter.evaluate(window, deviceState)) {
            QualityFilterDecision.Accepted -> ReadingWindowProcessingOutcome.Accepted(
                PersistedReadingCandidate(
                    latitude = window.latitude,
                    longitude = window.longitude,
                    roughnessRMS = window.roughnessRMS,
                    speedKmh = window.speedKmh,
                    headingDegrees = window.headingDegrees,
                    gpsAccuracyMeters = window.gpsAccuracyMeters,
                    isPothole = potholeCandidates.isNotEmpty(),
                    potholeMagnitudeG = potholeCandidates.maxOfOrNull { it.magnitudeG },
                    recordedAtEpochSeconds = window.startedAt + window.durationSeconds,
                    durationSeconds = window.durationSeconds,
                )
            )

            is QualityFilterDecision.Rejected -> ReadingWindowProcessingOutcome.Rejected(decision.reason)
        }
    }
}
